### ISSUE TITLE
Problems with new and old versions of V2 of ADMMPruner

### ISSUE THREAD

**Issue by typhoonlee** [microsoft/nni#4559]
**Describe the issue**:
I noticed that when using AMDD, there was a base_algo parameter in the original version，Can decide whether to prune weight, channel or filter：
`classnni.algorithms.compression.pytorch.pruning.ADMMPruner(model, config_list, trainer, criterion=CrossEntropyLoss(), num_iterations=30, epochs_per_iteration=5, row=0.0001, base_algo='l1')`

but in the pruningV2，there is no this parameter：
`classnni.algorithms.compression.v2.pytorch.pruning.ADMMPruner(model: torch.nn.modules.module.Module, config_list: List[Dict], trainer: Callable[[torch.nn.modules.module.Module, torch.optim.optimizer.Optimizer, Callable], None], traced_optimizer: nni.common.serializer.Traceable, criterion: Callable[[torch.Tensor, torch.Tensor], torch.Tensor], iterations: int, training_epochs: int)`
After use, the default pruning is the weight level, but I don't know where to modify it。

**Environment**:
- NNI version: V2.6
- Training service (local|remote|pai|aml|etc): local
- Client OS: linux
- Server OS (for remote mode only): 
- Python version: 3.8
- PyTorch/TensorFlow version: pytorch
- Is conda/virtualenv/venv used?: conda
- Is running in Docker?: no

**How to reproduce it?**:

**Comment by J-shang** [comment]
Hello @typhoonlee , different `base_algo` settings in admm v2 will support in `nni v2.8`.

For now, you can implement this by modifying the source code. Each basic pruner is the combination of `data_collector`, `metrics_calculator`, and `sparsity_allocator`. For example, if you want to prune the output channels with l1 norm just like `L1NormPruner`, modify these two lines:

https://github.com/microsoft/nni/blob/8b2eb425274cdb4537fbce4a315aec12a378d6db/nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py#L867

`self.metrics_calculator = NormMetricsCalculator(p=1, dim=0)`

https://github.com/microsoft/nni/blob/8b2eb425274cdb4537fbce4a315aec12a378d6db/nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py#L869

`self.sparsity_allocator = NormalSparsityAllocator(self, dim=0)`

`p=1` means use `L1 Norm`, `dim=0` means calculate metric on dim 0 of the weights, it is the output channels' dimension.

**Comment by typhoonlee** [comment]
> Hello @typhoonlee , different `base_algo` settings in admm v2 will support in `nni v2.8`.
> 
> For now, you can implement this by modifying the source code. Each basic pruner is the combination of `data_collector`, `metrics_calculator`, and `sparsity_allocator`. For example, if you want to prune the output channels with l1 norm just like `L1NormPruner`, modify these two lines:
> 
> https://github.com/microsoft/nni/blob/8b2eb425274cdb4537fbce4a315aec12a378d6db/nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py#L867
> 
> `self.metrics_calculator = NormMetricsCalculator(p=1, dim=0)`
> 
> https://github.com/microsoft/nni/blob/8b2eb425274cdb4537fbce4a315aec12a378d6db/nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py#L869
> 
> `self.sparsity_allocator = NormalSparsityAllocator(self, dim=0)`
> 
> `p=1` means use `L1 Norm`, `dim=0` means calculate metric on dim 0 of the weights, it is the output channels' dimension.

Thank you very much! Your answer effectively solved my problem.

**Comment by typhoonlee** [comment]
I encountered another problem, the code I used ADMM is as follows: But after getting the compressed small model, I performed finetune, the test result is very poor, not very common sense, because I only cut 4 channels：

```
def train_pipeline(root_path):
    # parse options, set distributed setting, set ramdom seed
    opt = parse_options(root_path, is_train=True)
    opt['root_path'] = root_path

    torch.backends.cudnn.benchmark = True
    # torch.backends.cudnn.deterministic = True

    # load resume states if necessary
    resume_state = load_resume_state(opt)
    # mkdir for experiments and logger
    if resume_state is None:
        make_exp_dirs(opt)
        if opt['logger'].get('use_tb_logger') and 'debug' not in opt['name'] and opt['rank'] == 0:
            mkdir_and_rename(osp.join(opt['root_path'], 'tb_logger', opt['name']))

    # initialize loggers
    logger, tb_logger = init_loggers(opt)

    # create train and validation dataloaders
    result = create_train_val_dataloader(opt, logger)
    train_loader, train_sampler, val_loader, total_epochs, total_iters = result

    def finetuner(model, optimizer_g, criterion):
        start_epoch = 0
        current_iter = 0
        # create message logger (formatted outputs)


        # dataloader prefetcher
        prefetch_mode = opt['datasets']['train'].get('prefetch_mode')
        if prefetch_mode is None or prefetch_mode == 'cpu':
            prefetcher = CPUPrefetcher(train_loader)
        elif prefetch_mode == 'cuda':
            prefetcher = CUDAPrefetcher(train_loader, opt)
            logger.info(f'Use {prefetch_mode} prefetch dataloader')
            if opt['datasets']['train'].get('pin_memory') is not True:
                raise ValueError('Please set pin_memory=True for CUDAPrefetcher.')
        else:
            raise ValueError(f'Wrong prefetch_mode {prefetch_mode}.' "Supported ones are: None, 'cuda', 'cpu'.")
        # training
        logger.info(f'Start training from epoch: {start_epoch}, iter: {current_iter}')
        data_time, iter_time = time.time(), time.time()  # 都是记录当前时刻
        start_time = time.time()

        def setup_optimizer(model, curerent_iter):
            train_opt = opt['train']
            optim_params = []
            for k, v in model.named_parameters():
                if v.requires_grad:
                    optim_params.append(v)
                else:
                    logger = get_root_logger()
                    logger.warning(f'Params {k} will not be optimized.')
            if 'type' in train_opt['optim_g']:
                optim_type = train_opt['optim_g'].pop('type')
            optimizer_g = torch.optim.Adam(optim_params, **train_opt['optim_g'])
            return optimizer_g

        def setup_schedulers(optimizer_g):
            train_opt = opt['train']
            if 'type' in train_opt['scheduler']:
                scheduler_type = train_opt['scheduler'].pop('type')
            else:
                scheduler_type = 'MultiStepLR'
            if scheduler_type in ['MultiStepLR', 'MultiStepRestartLR']:
                schedulers = lr_scheduler.MultiStepRestartLR(optimizer_g, **train_opt['scheduler'])
            return schedulers

        def update_learning_rate(current_iter, scheduler):
            if current_iter > 1:
                scheduler.step()

        def feed_data(data):
            lq = data['lq'].cuda()
            if 'gt' in data:
                gt = data['gt'].cuda()
            return lq, gt

        scheduler = setup_schedulers(optimizer_g)

        for epoch in range(start_epoch, (total_epochs//2 + 1)):
            train_sampler.set_epoch(epoch)
            prefetcher.reset()
            train_data = prefetcher.next()

            while train_data is not None:
                data_time = time.time() - data_time  # 加载数据的时间

                current_iter += 1
                if current_iter > total_iters:
                    break

                update_learning_rate(current_iter, scheduler)
                optimizer_g.zero_grad()
                lq, gt = feed_data(train_data)
                output = model(lq)
                # model.optimize_parameters(current_iter)  # 原始代码
                l_total = 0
                loss_dict = OrderedDict()
                l_pix = criterion(output, gt)
                l_total += l_pix
                loss_dict['l_pix'] = l_pix
                l_total.backward()
                optimizer_g.step()
                iter_time = time.time() - iter_time  # 每个iter的时间
                # log

                data_time = time.time()
                iter_time = time.time()
                train_data = prefetcher.next()

        consumed_time = str(datetime.timedelta(seconds=int(time.time() - start_time)))
        logger.info(f'End of training. Time consumed: {consumed_time}')
        logger.info('Save the latest model.')
    # create model
    if resume_state:  # resume training
        model = build_model(opt)
        model.resume_training(resume_state)  # handle optimizers and schedulers
        logger.info(f"Resuming training from epoch: {resume_state['epoch']}, " f"iter: {resume_state['iter']}.")
        start_epoch = resume_state['epoch']
        current_iter = resume_state['iter']
    else:
        model = build_model(opt)
        start_epoch = 0
        current_iter = 0

    # create message logger (formatted outputs)
    msg_logger = MessageLogger(opt, current_iter, tb_logger)

    # dataloader prefetcher
    prefetch_mode = opt['datasets']['train'].get('prefetch_mode')
    if prefetch_mode is None or prefetch_mode == 'cpu':
        prefetcher = CPUPrefetcher(train_loader)
    elif prefetch_mode == 'cuda':
        prefetcher = CUDAPrefetcher(train_loader, opt)
        logger.info(f'Use {prefetch_mode} prefetch dataloader')
        if opt['datasets']['train'].get('pin_memory') is not True:
            raise ValueError('Please set pin_memory=True for CUDAPrefetcher.')
    else:
        raise ValueError(f'Wrong prefetch_mode {prefetch_mode}.' "Supported ones are: None, 'cuda', 'cpu'.")

    # training
    logger.info(f'Start training from epoch: {start_epoch}, iter: {current_iter}')
    data_time, iter_time = time.time(), time.time()
    start_time = time.time()

    for epoch in range(start_epoch, total_epochs + 1):
        train_sampler.set_epoch(epoch)
        prefetcher.reset()
        train_data = prefetcher.next()

        while train_data is not None:
            data_time = time.time() - data_time

            current_iter += 1
            if current_iter > total_iters:
                break

            # Manual manual pruning
            if opt['Manuallyprune']:
                if current_iter == 1:
                    config_list = [{'op_types': ['Conv2d'], 'sparsity_per_layer': 0.125}, {'exclude': True, 'op_names': ['attconv']}]
                    traced_optimizer = nni.trace(torch.optim.Adam)(model.net_g.parameters(), lr=0.01)
                    criterion = build_loss(opt['train']['pixel_opt'])
                    pruner = ADMMPruner(model.net_g, config_list, finetuner, traced_optimizer=traced_optimizer, criterion=criterion, iterations=2, training_epochs=2)
                    model.net_g,mask = pruner.compress()

                    pruner._unwrap_model()
                    m_speedup = ModelSpeedup(model.net_g, dummy_input=torch.rand(1, 1, 28, 28).cuda(),
                                             masks_file=mask)
                    m_speedup.speedup_model()
                    # torch.save(model.net_g.state_dict(), '/media/amlogic/ed753a73-367b-40e7-87ca-ea306da3b61b/BasicSR-master/experiments/220218_ADMMPruner_(32-28)_150k_iter2epoch1_NOretrain/models/afterPrune.pth')
                    # print(model.net_g)
                    model.setup_optimizers()


            # update learning rate
            model.update_learning_rate(current_iter, warmup_iter=opt['train'].get('warmup_iter', -1))
            # training
            model.feed_data(train_data)
            model.optimize_parameters(current_iter)
            iter_time = time.time() - iter_time
            # log
            if current_iter % opt['logger']['print_freq'] == 0:
                log_vars = {'epoch': epoch, 'iter': current_iter}
                log_vars.update({'lrs': model.get_current_learning_rate()})
                log_vars.update({'time': iter_time, 'data_time': data_time})
                log_vars.update(model.get_current_log())
                msg_logger(log_vars)

            # save models and training states
            if current_iter % opt['logger']['save_checkpoint_freq'] == 0:
                logger.info('Saving models and training states.')
                model.save(epoch, current_iter)

            # validation
            # if opt.get('val') is not None and (current_iter % opt['val']['val_freq'] == 0):
            #     model.validation(val_loader, current_iter, tb_logger, opt['val']['save_img'])

            data_time = time.time()
            iter_time = time.time()
            train_data = prefetcher.next()
        # end of iter

    # end of epoch

    consumed_time = str(datetime.timedelta(seconds=int(time.time() - start_time)))
    logger.info(f'End of training. Time consumed: {consumed_time}')
    logger.info('Save the latest model.')
    model.save(epoch=-1, current_iter=-1)  # -1 stands for the latest
    # if opt.get('val') is not None:
    #     model.validation(val_loader, current_iter, tb_logger, opt['val']['save_img'])
    if tb_logger:
        tb_logger.close()


if __name__ == '__main__':
    root_path = osp.abspath(osp.join(__file__, osp.pardir, osp.pardir))
    train_pipeline(root_path)
```
